src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -302,6 +302,11 @@ void BaseCouplingScheme::firstExchange()
 
     _timeWindows += 1; // increment window counter. If not converged, will be decremented again later.
 
+    //If preCICE has stopped before the end of the time window we have to duplicate the last available sample and put it at the end of the time window.
+    if (getNextTimeStepMaxSize() > math::NUMERICAL_ZERO_DIFFERENCE && hasTimeWindowSize()) {
+      addTimeStepAtWindowEnd();
+    }
```

Comment:
```suggestion
    //If preCICE has stopped before the end of the time window we have to duplicate the last available sample and put it at the end of the time window.
    if (getNextTimeStepMaxSize() > math::NUMERICAL_ZERO_DIFFERENCE) {
      addTimeStepAtWindowEnd();
    }
```
Is this additional condition really needed? Why?

Besides that: Readability has improved a lot :+1:

Reply:
hasTimeWindowSize() is needed there, since if the participant does not have a time window size we can't put a sample at the end of the time window. I will add a comment clarifying this.

Reply:
True. Did not think about this point :+1: